File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.cc
// 66분
#include <iostream>
#include <algorithm>
#include <cstring>
using namespace std;

int n;
int arr[21][21];
int maxx = -1;
int isvisited[21][21];

void up() {
	// 모든 col에 대해서 진행.
	memset(isvisited, 0, sizeof(isvisited));

	for (int j = 0; j < n; j++) {
		for (int i = 1; i < n; i++) {
			if (arr[i][j] == 0) continue;
			int k = i-1;
			while (1) {
				if (k == 0 && arr[k][j] == 0) {
					arr[k][j] = arr[i][j];
					arr[i][j] = 0;
					break;
				}
				else if (k != 0 && arr[k][j] == 0) {
					k--;
				}
				else if (arr[k][j] != 0) {
					if (arr[k][j] != arr[i][j]) {
						if (k + 1 == i) break; // 위치 그대로인 경우
						else {
							arr[k + 1][j] = arr[i][j];
							arr[i][j] = 0;
							break;
						}
					}
					else if (arr[k][j] == arr[i][j]) {
						if (isvisited[k][j] == 0) {
							arr[k][j] = arr[k][j] * 2;
							arr[i][j] = 0;
							isvisited[k][j] = 1;
							break;
						}
						else if (isvisited[k][j] = 1) {
							if (k + 1 == i) break;
							else {
								arr[k + 1][j] = arr[i][j];
								arr[i][j] = 0;
								break;
							}
						}
					}
				}
			}
		}
	}
}

void down() { // 아래로 내려야한다.
	memset(isvisited, 0, sizeof(isvisited));
	for (int j = 0; j < n; j++) {
		for (int i = n-2; i >= 0; i--) {
			if (arr[i][j] == 0) continue;

			int k = i + 1; 
			while (1) {
				if (k == n-1 && arr[k][j] == 0) {
					arr[k][j] = arr[i][j];
					arr[i][j] = 0;
					break;
				}
				else if (k != 0 && arr[k][j] == 0) {
					k++;
				}
				else if (arr[k][j] != 0) {
					if (arr[k][j] != arr[i][j]) {
						if (k - 1 == i) break;
						else {
							arr[k - 1][j] = arr[i][j];
							arr[i][j] = 0;
							break;
						}
					}
					else if (arr[k][j] == arr[i][j]) {
						if (isvisited[k][j] == 0) {
							arr[k][j] = arr[k][j] * 2;
							arr[i][j] = 0;
							isvisited[k][j] = 1;
							break;
						}
						else if (isvisited[k][j] = 1) {
							if (k - 1 == i) break;
							else {
								arr[k - 1][j] = arr[i][j];
								arr[i][j] = 0;
								break;
							}
						}
					}
				}
			}
		}
	}
}

void left() {
	memset(isvisited, 0, sizeof(isvisited));

	for (int i = 0; i < n; i++) {
		for (int j = 1; j < n; j++) {
			if (arr[i][j] == 0) continue;
			int k = j - 1; 
			while (1) {
				if (k == 0 && arr[i][k] == 0) {
					arr[i][k] = arr[i][j];
					arr[i][j] = 0;
					break;
				}
				else if (k != 0 && arr[i][k] == 0) {
					k--;
				}
				else if (arr[i][k] != 0) {
					if (arr[i][k] != arr[i][j]) {
						if (k + 1 == j) break;
						else {
							arr[i][k + 1] = arr[i][j];
							arr[i][j] = 0;
							break;
						}
					}
					else if (arr[i][k] == arr[i][j]) {
						if (isvisited[i][k] == 0) {
							arr[i][k] = arr[i][k] * 2;
							arr[i][j] = 0;
							isvisited[i][k] = 1;
							break;
						}
						else if (isvisited[i][k] == 1) {
							if (k + 1 == j) break;
							else {
								arr[i][k + 1] = arr[i][j];
								arr[i][j] = 0;
								break;
							}
						}
					}
				}
			}
		}
	}
}

void right() {
	memset(isvisited, 0, sizeof(isvisited));

	for (int i = 0; i < n; i++) {
		for (int j = n-2; j >= 0; j--) {
			if (arr[i][j] == 0) continue;

			int k = j + 1; 
			while (1) {
				if (k == n-1 && arr[i][k] == 0) {
					arr[i][k] = arr[i][j];
					arr[i][j] = 0;
					break;
				}
				else if (k != 0 && arr[i][k] == 0) {
					k++;
				}
				else if (arr[i][k] != 0) {
					if (arr[i][k] != arr[i][j]) {
						if (k - 1 == j) break;
						else {
							arr[i][k - 1] = arr[i][j];
							arr[i][j] = 0;
							break;
						}
					}
					else if (arr[i][k] == arr[i][j]) {
						if (isvisited[i][k] == 0) {
							arr[i][k] = arr[i][k] * 2;
							arr[i][j] = 0;
							isvisited[i][k] = 1;
							break;
						}
						else if (isvisited[i][k] == 1) {
							if (k - 1 == j) break;
							else {
								arr[i][k - 1] = arr[i][j];
								arr[i][j] = 0;
								break;
							}
						}
					}
				}
			}
		}
	}
}

void dfs(int cnt) {
	if (cnt == 6) {
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < n; j++) {
				maxx = max(maxx, arr[i][j]);
			}
		}
		return;
	}

	int tmparr[21][21];

	copy(&arr[0][0], &arr[0][0] + 21 * 21, &tmparr[0][0]);
	up();
	dfs(cnt + 1);
	copy(&tmparr[0][0], &tmparr[0][0] + 21 * 21, &arr[0][0]);

	copy(&arr[0][0], &arr[0][0] + 21 * 21, &tmparr[0][0]);
	down();
	dfs(cnt + 1);
	copy(&tmparr[0][0], &tmparr[0][0] + 21 * 21, &arr[0][0]);

	copy(&arr[0][0], &arr[0][0] + 21 * 21, &tmparr[0][0]);
	left();
	dfs(cnt + 1);
	copy(&tmparr[0][0], &tmparr[0][0] + 21 * 21, &arr[0][0]);

	copy(&arr[0][0], &arr[0][0] + 21 * 21, &tmparr[0][0]);
	right();
	dfs(cnt + 1);
	copy(&tmparr[0][0], &tmparr[0][0] + 21 * 21, &arr[0][0]);
}

int main() {
	cin >> n;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < n; j++) {
			cin >> arr[i][j];
		}
	}

	dfs(1);
	cout << maxx << '\n';
}

File: 백준/Gold/12100. 2048 （Easy）/2048 （Easy）.py
n = int(input().strip())

llist = [list(map(int, input().strip().split())) for _ in range(n)]
# 1. up, down, right, left 메서드 구현하기
res = 0

def copy_list(llist2): # llist2가 llist꺼를 저장함함
    global n
    for i in range(n):
        for j in range(n):
            llist2[i][j] = llist[i][j]

def recover_list(llist2): # llist가 llist2 내용물을 가져옴
    global n
    for i in range(n):
        for j in range(n):
            llist[i][j] = llist2[i][j]

def update_res():
    global n, res
    for i in range(n):
        for j in range(n):
            res = max(res, llist[i][j])

def dfs(cnt):
    global res, n
    if cnt == 5:
        # 25번 돌면서 res 갱신
        update_res()
        return

    llist2 = [[0] * n for _ in range(n)]
    # llist2 = llist하기
    copy_list(llist2)
    # print("uuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuu")
    up()
    dfs(cnt+1)
    # llist를 llist2로 변경하기기
    recover_list(llist2)

    copy_list(llist2)
    # print("dddddddddddddddddddddd")
    down()
    dfs(cnt+1)
    recover_list(llist2)

    copy_list(llist2)
    # print("rrrrrrrrrrrrrrrrrrrrrrrrr")
    right()
    dfs(cnt+1)
    recover_list(llist2)
    
    copy_list(llist2)
    # print("llllllllllllllllllllllllllll")
    left()
    dfs(cnt+1)
    recover_list(llist2)

def up():
    global n
    for j in range(n):
        i = 1
        is_ch = 0
        while i < n:
            # print(i, "최초 i")
            # printt()
            # print("프린트 함 해야지지")
            if llist[i][j] == 0:
                i+=1
                continue
            elif llist[i][j] != 0:
                tmp_i = i-1
                while True:
                    if tmp_i == -1: break
                    if llist[tmp_i][j] != 0: break
                    tmp_i-=1
                # print("현재 i는", i, "현재 tmpi는", tmp_i)
                if tmp_i == -1: # tmp_i가 -1 이라는 의미 = 젤 위로 올리면됨
                    llist[0][j] = llist[i][j]
                    llist[i][j] = 0
                else: # tmp_i가 -1이 아니라는거는 0이 아닌거를 만났다는것
                    if llist[tmp_i][j] == llist[i][j] and is_ch == 0:
                        llist[tmp_i][j] = llist[tmp_i][j]*2
                        llist[i][j] = 0
                        is_ch = 1
                    elif llist[tmp_i][j] != llist[i][j] or is_ch == 1:
                        is_ch = 0
                        if tmp_i+1 != i:
                            llist[tmp_i+1][j] = llist[i][j]
                            llist[i][j] = 0
                i+=1
                # print(i)
def down():
    global n
    for j in range(n):
        i = n-2
        is_ch = 0
        while i >= 0:
            # print(i, "최초 i")
            # printt()
            # print("프린트 함 해야지지")
            if llist[i][j] == 0:
                i-=1
                continue
            elif llist[i][j] != 0:
                tmp_i = i+1
                while True:
                    if tmp_i == n: break
                    if llist[tmp_i][j] != 0: break
                    tmp_i+=1
                # print("현재 i는", i, "현재 tmpi는", tmp_i)
                if tmp_i == n: # tmp_i가 -1 이라는 의미 = 젤 위로 올리면됨
                    llist[n-1][j] = llist[i][j]
                    llist[i][j] = 0
                else: # tmp_i가 -1이 아니라는거는 0이 아닌거를 만났다는것
                    if llist[tmp_i][j] == llist[i][j] and is_ch == 0:
                        llist[tmp_i][j] = llist[tmp_i][j]*2
                        llist[i][j] = 0
                        is_ch = 1
                    elif llist[tmp_i][j] != llist[i][j] or is_ch == 1:
                        is_ch = 0
                        if tmp_i-1 != i:
                            llist[tmp_i-1][j] = llist[i][j]
                            llist[i][j] = 0
                i-=1
                # print(i)

def right():
    global n
    for i in range(n):
        j = n-2
        is_ch = 0
        while j >= 0:
            if llist[i][j] == 0:
                j-=1
                continue
            elif llist[i][j] != 0:
                tmp_j = j+1
                while True:
                    if tmp_j == n: break
                    if llist[i][tmp_j] != 0: break
                    tmp_j+=1
                if tmp_j == n: # tmp_i가 -1 이라는 의미 = 젤 위로 올리면됨
                    llist[i][n-1] = llist[i][j]
                    llist[i][j] = 0
                else: # tmp_i가 -1이 아니라는거는 0이 아닌거를 만났다는것
                    if llist[i][tmp_j] == llist[i][j] and is_ch == 0:
                        llist[i][tmp_j] = llist[i][tmp_j]*2
                        llist[i][j] = 0
                        is_ch = 1
                    elif llist[i][tmp_j] != llist[i][j] or is_ch == 1:
                        is_ch = 0
                        if tmp_j-1 !=j:
                            llist[i][tmp_j-1] = llist[i][j]
                            llist[i][j] = 0
                j-=1
                # print(i)


def left():
    global n
    for i in range(n):
        j = 1
        is_ch = 0
        while j < n:
            if llist[i][j] == 0:
                j+=1
                continue
            elif llist[i][j] != 0:
                tmp_j = j-1
                while True:
                    if tmp_j == -1: break
                    if llist[i][tmp_j] != 0: break
                    tmp_j-=1
                if tmp_j == -1: # tmp_i가 -1 이라는 의미 = 젤 위로 올리면됨
                    llist[i][0] = llist[i][j]
                    llist[i][j] = 0
                else: # tmp_i가 -1이 아니라는거는 0이 아닌거를 만났다는것
                    if llist[i][tmp_j] == llist[i][j] and is_ch == 0:
                        llist[i][tmp_j] = llist[i][tmp_j]*2
                        llist[i][j] = 0
                        is_ch = 1
                    elif llist[i][tmp_j] != llist[i][j] or is_ch == 1:
                        is_ch = 0
                        if tmp_j+1 != j:
                            llist[i][tmp_j+1] = llist[i][j]
                            llist[i][j] = 0
                j+=1


def printt():
    global n
    print("*************^*^**")
    for i in range(n):
        for j in range(n):
            print(llist[i][j], end = " ")
        print()
    print("************")

dfs(0)
print(res)

'''
4
8 8 8 8
4 4 4 4
2 2 2 2
2 2 2 2

4
4 4 4 4
0 0 0 0
2 2 2 0
2 2 2 4

4
2 2 2 2
2 2 2 2
4 4 4 4
8 8 8 8
'''
